File: team-app-backend/src/main/java/com/cillu/adapters/HttpHandler.java
package com.cillu.adapters;

import com.cillu.core.domain.Player;
import com.cillu.core.domain.Team;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import com.cillu.ports.TeamServicePort;

import java.util.List;

@RestController
public class HttpHandler {

    private TeamServicePort teamServicePort;

    public HttpHandler(TeamServicePort teamServicePort){
        this.teamServicePort = teamServicePort;
    }

    @GetMapping("/api/v1/teams")
    @ResponseBody
    public List<Team> listTeams() throws Exception {
        return teamServicePort.listTeams();
    }

    @PostMapping("/api/v1/teams")
    @ResponseBody
    public Team createTeam(@RequestBody Team team) throws Exception{
        return teamServicePort.createNewTeam(team);
    }

    @GetMapping("/api/v1/teams/{teamID}/players")
    @ResponseBody
    public List<Player> getTeamPlayers(@PathVariable int teamID) throws Exception{
        return teamServicePort.listTeamPlayers(teamID);
    }

    @PostMapping("/api/v1/teams/{teamID}/invite-player")
    @ResponseBody
    public Player invitePlayer(@RequestBody Player player, @PathVariable int teamID) throws Exception{
        return teamServicePort.invitePlayer(player);
    }

}

File: adapters/http_handler.go
package adapters

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/thecillu/exagonal-architecture/core/domain"
	"github.com/thecillu/exagonal-architecture/ports"
)

type httpHandler struct {
	service ports.TeamServicePort
}

func NewHttpHandler(service ports.TeamServicePort) *httpHandler {
	return &httpHandler{
		service: service,
	}
}

func (hdl *httpHandler) RetrieveTeams(c *gin.Context) {
	teams, err := hdl.service.ListTeams()
	if err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, teams)
}

func (hdl *httpHandler) CreateNewTeam(c *gin.Context) {
	var input domain.Team
	if err := c.ShouldBindJSON(&input); err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
		return
	}
	team, err := hdl.service.CreateNewTeam(input)
	if err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, team)
}

func (hdl *httpHandler) ListTeamPlayers(c *gin.Context) {
	team_id, err := strconv.ParseInt(c.Param("team_id"), 10, 64)
	players, err := hdl.service.ListTeamPlayers(team_id)
	if err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, players)
}

func (hdl *httpHandler) InvitePlayer(c *gin.Context) {
	var input domain.Player
	if err := c.ShouldBindJSON(&input); err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
		return
	}
	player, err := hdl.service.InvitePlayer(input)
	if err != nil {
		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, player)
}
